agent/wsclient/decode.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2014-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Package wsclient wraps the generated aws-sdk-go client to provide marshalling
+// and unmarshalling of data over a websocket connection in the format expected
+// by backend. It allows for bidirectional communication and acts as both a
+// client-and-server in terms of requests, but only as a client in terms of
+// connecting.
+package wsclient
+
+import (
+	"bytes"
+	"encoding/json"
+	"reflect"
+
+	"github.com/awslabs/aws-sdk-go/internal/protocol/json/jsonutil"
+)
+
+// DecodeData decodes a raw message into its type. E.g. An ACS message of the
+// form {"type":"FooMessage","message":{"foo":1}} will be decoded into the
+// corresponding *ecsacs.FooMessage type. The type string, "FooMessage", will
+// also be returned as a convenience.
+func DecodeData(data []byte, dec TypeDecoder) (interface{}, string, error) {
```

Comment:
Is this ever called outside the `wsclient` package?  If not, I'd prefer it be made private.